### DISCUSSION TITLE
ESP32-CAM: cannot get sdcard operation to work

### DISCUSSION THREAD

**Discussion by Flipje1955** [micropython/micropython#13407] [ESP32] [answered]
getting first hand-on experience with this mpu and got the primary goal for buing it to work (the camera) to work and managed to send the pictures to a simple webserver. As next step, I want to save the pictures using the sdcard reader/writer this mcu has. 
Using what I hope are the correct  settings for the hw SPI with id = 1 (mosi: GPIO 13, miso: GPIO 12, clk: GPIO 14 and cs: GPIO 15), it keeps failing with "OSError: no sdcard".  Tried some different pins but most of the pins are already "attached" to the camera.  Is there anybody who has successfully controlled the onboard sdcard interface and can offer advice ?

- Cards are the same I successfully use for rp2040 Pico / sdcard operations;  the driver is the official mp sdcard module.
- The ESP32-CAM is a special version of the ESP32 and there are many privately build firmware versions around (The one I use says: MicroPython v1.21.0 on 2023-10-13); it is not clear to me if and then how they relate to official mp releases. nb. There is a pull request to create official MP firmware specifically for the ESP32-CAM mpu; I hope that request will make it.
- There is a known limitation for the ESP32-CAM sdcards to not exceed 4GB but from Google searches I understand that that implies only 4GB is accessible and not that bigger cards would be rejected. Both 16G and 32G (FAT32 formatted) cards fail. Regrettably, I do not have any less_than 4G sized cards to try.

**Comment by shariltumin** [comment, marked as answer]
The sdcard onboard of esp32-cam (AIThinker) is wired as sdmmc device.

Using SPI to access the sdcard will not work as you mention. You have to use ```sd=machine.SDCard()```.

```py
>>> import machine, os
>>> sd=machine.SDCard()
>>> sd.info()
(4016046080, 512)
>>> os.mount(sd, "/sd")
>>> fn = '/sd/test'
>>> f = open(fn, 'w')
>>> msg = 'hello world!'
>>> w=f.write(msg)
>>> w == len(msg)
True
>>> f.flush()
>>> f.close()
>>> f = open(fn)
>>> ms = f.read()
>>> ms == msg
True
>>> f.close()
>>> os.umount("/sd")
>>> sd.deinit()
```

Using an SD card with a capacity greater than 4GB will cause the system to hang.

The firmware used in the test is MicroPython v1.20.0-206-g33b403dfb compiled with idf4.4.6 which support OV2640.

**Reply by Flipje1955** [reply]
thank you very much for your solution and quick response ! .... that also works "my" firmware (btw also with 16GB and 32GB cards !)
